hash/hash_comparator.go

```diff
@@ -162,11 +162,16 @@ func CompareMD5Crypt(_ context.Context, password []byte, hash []byte) error {
 	return compareCryptHelper(password, string(hash))
 }
 
-func CompareBcrypt(_ context.Context, password []byte, hash []byte) error {
+func CompareBcrypt(ctx context.Context, password, hash []byte) error {
 	if err := validateBcryptPasswordLength(password); err != nil {
 		return err
 	}
 
+	// ensure that the context is not canceled before doing the heavy lifting
+	if ctx.Err() != nil {
```

Comment:
Does this actually do anything? Isn't the hashing happening sequentially?

Reply:
I went ahead and added the context check in every place before we do any hash computation, as that's the best we can do. The comparison typically only happens during login, so there should not be more than one hash being computed, however I figured that it doesn't hurt to check there as well.